### ISSUE TITLE
simulation ambiguous between the FBPIC and Warp code

### ISSUE THREAD

**Issue by QJohn2017** [fbpic/fbpic#368]
Hi, @RemiLehe , @soerenjalas 

I used the same plasma parameters and laser parameters in FBPIC code and [Warp code ](https://bitbucket.org/berkeleylab/warp/src/master/). And in Warp code, I also used the  quasi-cylindrical geometry mode like the FBPIC. the input parameters are the same with that in #362.

The simulation results  by the FBPIC has showed in #366 by @soerenjalas , it can be seen that the injected electron beams in that figures are good and do not dephase yet, and its energy is about 1.5 GeV. 

But when I used the same conditions in Warp code, the results are below:
![11](https://user-images.githubusercontent.com/31507149/58805399-95c41880-8646-11e9-897d-605feb9bc31b.jpg)

It can be seen that the injected electron beams here are not good and its energy is only about 600 MeV. So, i don't know which PIC code is correct to simulate my case , can you help me ? Thank you !

**Comment by RemiLehe** [comment]
Hi @QJohn2017.

The key difference here is that Warp uses a finite-difference scheme for solving the Maxwell equations, whereas FBPIC uses a spectral solver. 

In general, finite-difference schemes are much more likely to have signifiant **numerical dispersion**. In typical laser-wakefield simulations, numerical dispersion can cause the laser driver to travel at lower speed **in the simulation** than it would physically. This can cause the electron beam to artificially dephase with respect to the laser, in simulations, and I think that this is what you are seeing in the Warp simulation. For more details, I would strongly recommend to read the [original paper on FBPIC](https://www.sciencedirect.com/science/article/pii/S0010465516300224) and in particular sections 3 and 4, which highlight differences between the spectral and finite-difference solvers.

So in summary, I think that, in your particular case, a spectral solver is more well-suited than a finite-difference solver (because the reduced numerical dispersion), and thus the FBPIC results are probably more correct than the Warp results.

Does this answer your question? If so, could you close the issue?

**Comment by QJohn2017** [comment]
Sorry, I'm not entirely convinced by your answer, but I take a cursory look at the original paper on FBPIC , it showed that the methods to eliminate the spurious numerical dispersion in  finite-difference scheme are finer grid in z direction (which is very computational expensive ) or coarser grid in r direction (which delta r equals to 5 delta z).  In my warp simulation, the delta r equals to 10 times of delta z. So, I want to use a coarser gird in r direction to simulate my case by setting delta r equals to 5 times of delta z in Warp and see whether the simulation results are improved or not. Do you think this is ok ? Or, do you have a better way to help me prove the simulation results?

**Comment by RemiLehe** [comment]
Looking at the parameters that you posted in #362, it seems that you have `dz = 5e-8` and `dr = 2.67e-7` (from `rmax = 80e-6` and `Nr=300`)
Assuming that you used the same `dz` and `dr` for your first Warp simulation, then yes, the Warp simulation results will probably be better (in the sense that you will see **less** dephasing), by using a **coarser** `dr`, for instance `dr = 4.e-7` (e.g. using the same `rmax` but `Nr=200`).

**Comment by QJohn2017** [comment]
yes, I used the same `dz` and `dr` (from` rmax = 80e-6 and Nr=300, Nz=1680, zmin = -64um, zmmax=20um`) in warp simulation before, which is the same with that in FBPIC simulation. So, according to your suggestion, I will change the `Nr` to 200 and use the same `rmax` in Warp code , which means that the `dr = 8*dz`, i will try again and let you know as soon as I have the results.

**Comment by QJohn2017** [comment]
@RemiLehe , In my latest simulation with warp code,I don't change the `dr`, but only change the field solver from `Yee solver` to `Lehe solver` and the results are the same with FBPIC code, the last results of warp code are below:
![14](https://user-images.githubusercontent.com/31507149/58968047-eb83f680-87e7-11e9-8485-55ac95bf5863.jpg)
It can be seen that this results are almost the same with that in FBPIC in #366. So, you are right ! Thak you !

PS: it is very confusion that under the same condition, the run time  in `Warp` code only takes 12 hours, but the run time in FBPIC code takes nearly 4.5 days !

**Comment by RemiLehe** [comment]
@QJohn2017 Yes, using this solver in Warp is a good option to reduce the amount of numerical dispersion. However, note that the "Lehe solver" is not strictly dispersion-free either (although it does have much lower spurious numerical dispersion than the Yee solver; more details in [this paper](https://journals.aps.org/prab/abstract/10.1103/PhysRevSTAB.16.021301)).

Regarding the difference in speed: in principle, it is expected that FBPIC would somewhat slower than Warp (because of the computational cost of the Hankel transform), but definitely not at the 10x level that you are observing.
I think that the reason why you are observing such a large difference in performance is probably because of the different way in which Warp and FBPIC are parallelized: Warp has pure MPI parallelization, whereas FBPIC has a hybrid MPI+multi-threading parallelization. 
It seems that the computing environment that you are using is not able to efficiently use multi-threading (e.g. it might be that, for some reason, all threads are trying to run on the same, single physical core). I think this could explain the difference in performance (on CPU) between your computing environment and @MKirchen's environment (#357).

Since it seems that your original question was answered, I will close this issue.
